File: results/RQ1/05_baseline_qualification/units/p05_brick_breaker_game/cpp/accepted/repository/src/Level.cpp
#include "Level.h"

namespace breakout {

Level::Level() = default;

void Level::generateGrid(int rows, int cols, Vector2 origin, Vector2 cellSize, int hp) {
    bricks.clear();
    
    for (int row = 0; row < rows; ++row) {
        for (int col = 0; col < cols; ++col) {
            Vector2 position(
                origin.x + col * cellSize.x,
                origin.y + row * cellSize.y
            );
            bricks.emplace_back(position, cellSize.x, cellSize.y, hp);
        }
    }
}

std::size_t Level::remaining() const {
    std::size_t count = 0;
    for (const auto& brick : bricks) {
        if (brick.isAlive()) {
            ++count;
        }
    }
    return count;
}

} // namespace breakout

File: results/RQ1/05_baseline_qualification/units/p05_brick_breaker_game/python/attempts/0002_seed_3996841867/p05_brick_breaker_game_python/p05_brick_breaker_game_python/brick_breaker/level.py
"""Level implementation for the brick breaker game.

This module defines the Level class which manages a collection of bricks
and provides functionality to create a grid layout of bricks.
"""

from .config import GameConfig
from .brick import Brick


class Level:
    """Represents a level in the brick breaker game, containing bricks."""

    def __init__(self, bricks: list[Brick]):
        """Initialize a level with a list of bricks.

        Args:
            bricks: List of Brick objects making up the level.
        """
        self.bricks = bricks

    def remaining(self) -> int:
        """Count the number of bricks that are still alive.

        Returns:
            The number of bricks with durability > 0.
        """
        return sum(1 for brick in self.bricks if brick.alive())


def make_grid_level(config: GameConfig) -> Level:
    """Create a grid of bricks based on configuration parameters.

    Args:
        config: GameConfig object containing level layout parameters.

    Returns:
        A Level object populated with bricks arranged in a grid pattern.
    """
    # Calculate brick width if not explicitly provided
    if config.brick_width is None:
        total_width = config.brick_cols * config.brick_width + (config.brick_cols - 1) * config.brick_padding
        brick_width = (config.width - total_width) / config.brick_cols
    else:
        brick_width = config.brick_width

    # Create bricks in a grid pattern
    bricks = []
    for row in range(config.brick_rows):
        for col in range(config.brick_cols):
            # Calculate x and y position for this brick
            x = col * (brick_width + config.brick_padding) + config.brick_padding
            y = row * (config.brick_height + config.brick_padding) + config.top_margin
            
            # Create brick and add to list
            brick = Brick(x, y, brick_width, config.brick_height)
            bricks.append(brick)

    return Level(bricks)